feat: add terminal UI to generate compose files

shantanoo-desai opened the pull request:
provide a terminal based UI using whiptail to generate a
custom `docker-compose.yml` file. The bash script collects the
necessary selections from a defined Usage Flow and generates the
compose file by calling the `make compose <selection>` command.
Aforementioned Usage Flow is documented in #225 

Signed-off-by: Shantanoo 'Shan' Desai <shantanoo.desai@gmail.com>

<!-- Expected Commit Message Description (imported automatically by GitHub) -->
<!-- Must conform to [conventional commits guidelines](https://github.com/edgexfoundry/edgex-compose/blob/main/.github/Contributing.md) -->
<!-- Expected Commit message must contain Closes/Fixes #IssueNumber statement when there is a related issue -->

<!-- Add additional detailed description of need for change if no related issue -->

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-compose/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>


## Testing Instructions
<!-- How can the reviewers test your change? -->
```bash
cd compose-builder/
chmod +x tui-generator.sh
./tui-generator.sh
```
### Tested with:
__OS/Distro__: GNU/Linux Manjaro Rolling
__whiptail version__: `whiptail (newt): 0.52.21`
__GNU make__: 4.3
__docker version__: Docker version 20.10.12, build e91ed5707e
__docker-compose version__: docker-compose version 1.29.2

## Docs
Current PR under review and documentation follows once the team decides upon a fixed flow of UI and whether the code needs refactoring. For actual discussion refer to #225

lenny-intel commented:
@shantanoo-desai , playing around with this and a couple things that I think might improve user experience.

1. Have the option at the very end to `generate` (make gen) or `generate and run` (make run)
2. Have the ability to go back to the previous screen. I found a couple time I hit enter and then wanted to go back.

lenny-intel commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,258 @@
+#!/usr/bin/env bash
+
+# /*******************************************************************************
+#  * Copyright 2022 
```
Copyright needs an entity. Your company?

shantanoo-desai replied:
I am committing as a stand-alone contributor here. Should I then add my own name or are there different policies for contribution?

shantanoo-desai replied:
Have added my e-mail address, as a personal contribution

shantanoo-desai commented:
@lenny-intel  thanks for the review and feedback:

> 1. Have the option at the very end to `generate` (make gen) or `generate and run` (make run)

I think this is possible! I will try redrafting a usage flow with a dialog box at the end of the setup

> 2. Have the ability to go back to the previous screen. I found a couple time I hit enter and then wanted to go back.

This is a bit tricky. However instead of `Cancel` I can rename the button as `Back` and hopefully it might work out. I will get back to you on this.

shantanoo-desai commented:
@lenny-intel I have introduced the final step of `Generate File` and `Generate File and Run`. I was not successful in creating a `Back` to previous menu as `whiptail` with `bash` does not provide syntax like `goto`. This will only lead to redundant if else blocks which might become messy and won't serve as a solution but rather a hack. I have added a bit more verbose messages about the <kbd>Enter</kbd> key. Please let me know if you have some more feedback

shantanoo-desai commented:
@lenny-intel any updates or opinions on the changes? Also I am not sure why the tests have failed I have rebased the PR.

cloudxxx8 commented:
recheck

lenny-intel commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,293 @@
+        arglist+=("${additionalOptsDesc[$index]}")
+        arglist+=("OFF") # Nothing is selected by-default
+    done
+    SELECTED_OTHERS+=$(/usr/bin/whiptail --title "App Services" \
```
Don't assume where whiptail is install. This fails for me. Just assume it is in the user's path. You may what to add a check that it is available and fail out if not.
```suggestion
    SELECTED_OTHERS+=$(whiptail --title "App Services" \
```

shantanoo-desai replied:
This is resolved now by initially checking whether `whiptail` exists, if not exit the script. If it does exist the path to the binary will be stored in a variable and then resolved at all places where `whiptail` was initially used.

lenny-intel commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,293 @@
+#    ARM64 Question Display Function                               #
+####################################################################
+function displayArm64() {
+    message="Would you like to use ARM64 Images to generate the Compose file?"
+
+    /usr/bin/whiptail --title "Images Architecture" --yesno "$message" 8 78
```
Reverse the logic here since `Not ARM64` will be the most common answer so should be the default.

shantanoo-desai replied:
This is resolved by setting `--defaultno` on the Yes / No Message box. By default the box will have the "cursor" set on No.

lenny-intel commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,293 @@
+####################################################################
+function devServiceOption() {
+    message="Available Device Services to include in your compose file.\n
+            Press <SPACEBAR> to Select \n
+            Press <ENTER> to Skip \n
```
This text is not being show for me.
![image](https://user-images.githubusercontent.com/44779287/164527886-6e4d0896-ccf0-4d65-ab3d-fe2c743e5222.png)

lenny-intel replied:
It is for the App Services. I suspect it is a issue with screen space available in the small box. Can you make the box larger?

lenny-intel reviewed (changes requested):
I would like to **CTRL-C** out  to `Cancel` but can not. Had to go all the way through. Can you add `Cancel` via  **CTRL-C**

shantanoo-desai commented:
@lenny-intel I wasn't successful in trying to trap either <kbd>CTRL</kbd> + <kbd>C</kbd> or <kbd>ESC</kbd> within the script. There is unfortunately no solution I am aware of that is working although the `man` pages mention <kbd>ESC</kbd> and [This AskUbuntu Query](https://askubuntu.com/questions/1137617/how-to-quit-whiptail-gauge) mentioning about using `export TERM=linux`. None of the solutions are able to trap the signals to exit.

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+
+if [[ $? == 0 ]]; then
+    echo "Generating Compose file...\n"
+    $MAKE gen $SELECTED_OTHERS $SELECTED_DEVSERVICES $SELECTED_APPSERVICES $SELECTED_BUS
```
Since these are arrays, you should reference them as such:

```suggestion
    $MAKE gen "${SELECTED_OTHERS[@]}" "${SELECTED_DEVSERVICES[@]}" "${SELECTED_APPSERVICES[@]}" "${SELECTED_BUS[@]}"
```

ajcasagrande replied:
Now that I look closer, they are not actually arrays, they were just defined at the top as such, so doing it this way will fail. They should either be redefined as strings or read into an array from whiptail's response.

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+    echo "Generating Compose file and Running the Stack....\n"
+    $MAKE gen $SELECTED_OTHERS $SELECTED_DEVSERVICES $SELECTED_APPSERVICES $SELECTED_BUS
+    $MAKE run
+fi
```
Missing newline/blank line at the end of the file.

Press the enter key after this line to end the file with a newline (`\n`)

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+    $MAKE gen $SELECTED_OTHERS $SELECTED_DEVSERVICES $SELECTED_APPSERVICES $SELECTED_BUS
+else
+    echo "Generating Compose file and Running the Stack....\n"
+    $MAKE gen $SELECTED_OTHERS $SELECTED_DEVSERVICES $SELECTED_APPSERVICES $SELECTED_BUS
+    $MAKE run
```
In the makefile, `run` depends on `gen` so your run command may actually be ignoring your first gen command.

```suggestion
    $MAKE run "${SELECTED_OTHERS[@]}" "${SELECTED_DEVSERVICES[@]}" "${SELECTED_APPSERVICES[@]}" "${SELECTED_BUS[@]}"
```

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+}
+
+## Step - 1: Start by Asking about whether Images should be pulled for ARM64?
+displayArm64
+if [[ $? == 0 ]]; then
```
check the result of the call directly:
```suggestion
if displayArm64; then
```

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+fi
+
+## Step - 2: Ask whether a non-secure Configuration is needed
+displayNoSecty
+if [[ $? == 0 ]]; then
```
check the result of the call directly:
```suggestion
if displayNoSecty; then
```

shantanoo-desai replied:
Here, I think, there is a different pattern `no-secty` should only be set when the user selects `No` or `Non-Secure`.

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+#    Message Bus Options Display Function                          #
+####################################################################
+function msgBusOption() {
+    message="Available Message Buses to replace the default one.\n
```
I would probably use the word `Alternative` like: `Alternative Message Buses to replace the default one`

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+#    Additional Services Options Display Function                  #
+####################################################################
+function additionalServiceOption() {
+    message="Some Additional Services are available that can also be included. \n
```
Anywhere you have defined the `message` here like this, you should probably make it a `local` like so:
```shell
local message="...."
```

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+
+    # Generate a Specific form of Array required by whiptail checklist
+    # FORMAT: "<INDEX> <DESCRIPTION> <OFF>"
+    arglist=()
```
you should probably make all of these `arglist` defines `local`:
```suggestion
    local arglist=()
```

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+## Step - 1: Start by Asking about whether Images should be pulled for ARM64?
+displayArm64
+if [[ $? == 0 ]]; then
+    SELECTED_OTHERS+=("arm64 ")
```
Instead of lumping these in with the others, I would probably just have a specific one:
```shell
ARM64_OPT=""
if displayArm64; then
    ARM64_OPT="arm64"
fi
```

then, in your make gen, just reference it with `"${ARM64_OPT}"`. If it is blank, nothing with be added to the command.

shantanoo-desai replied:
I tried your suggestion however, one problem that occurs when using `"${SECURITY_OPT}"` with the following:

    make gen "${ARM_64_OPT}" "${SECURITY_OPT}"

is the fact that when both are left blank (not-selected) then `make` throws the following error:

    make: *** empty string invalid as file name.  Stop.

I suspect that two empty spaces leads before this leads to such a problem. A potential solution is just to resolve the variables without the double quotes `"`

    make gen ${ARM64_OPT} ${SECURITY_OPT}

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+## Step - 2: Ask whether a non-secure Configuration is needed
+displayNoSecty
+if [[ $? == 0 ]]; then
+    SELECTED_OTHERS+=("no-secty ")
```
Same as the ARM64 one, i would make a `SECURITY_OPT`

Instead of lumping these in with the others, I would probably just have a specific one:
```shell
SECURITY_OPT=""
if displayNoSecty; then
    SECURITY_OPT="no-secty"
fi
```

then, in your make gen, just reference it with `"${SECURITY_OPT}"`. If it is blank, nothing with be added to the command.

shantanoo-desai replied:
@ajcasagrande 
I tried your suggestion however, one problem that occurs when using `"${SECURITY_OPT}"` with the following:

    make gen "${ARM_64_OPT}" "${SECURITY_OPT}"

is the fact that when both are left blank (not-selected) then `make` throws the following error:

    make: *** empty string invalid as file name.  Stop.

I suspect that two empty spaces leads before this leads to such a problem. A potential solution is just to resolve the variables without the double quotes `"`

    make gen ${ARM64_OPT} ${SECURITY_OPT}

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+#    No-Security Question Display Function                         #
+####################################################################
+function displayNoSecty() {
+    message="Would you like to generate a non-secure configuration of the Compose file?"
```
I feel this question is a little backwards, or worded stangely. Maybe instead of yes-no it should be `Secure` and `Non-Secure` for the buttons kind of like the final step?

shantanoo-desai replied:
@ajcasagrande How about refactoring the question to

> Choose a Security configuration for EdgeX?

where the options would be either `Secure` or `Non-Secure`

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+    arglist=()
+    for index in "${deviceServices[@]}";
+    do
+        arglist+=("$index")
```
same here, simplify to 1 line

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+## Step -6: Ask for Additional Options
+additionalServiceOption
+
+# echo "additional services selected: ${SELECTED_OTHERS}" ## DEBUG
+# echo "device services selected: ${SELECTED_DEVSERVICES}" ## DEBUG
+# echo "app services selected: ${SELECTED_APPSERVICES}" ## DEBUG
+# echo "selected message bus: ${SELECTED_BUS}" ## DEBUG
```
commented out code should be removed before  merged

ajcasagrande commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+
+
+## Available Device Services in EdgeX-Foundry
+deviceServices=(
+    "ds-bacnet"
+    "ds-camera"
+    "ds-grove"
+    "ds-modbus"
+    "ds-mqtt"
+    "ds-rest"
+    "ds-snmp"
+    "ds-virtual"
+    "ds-coap"
+    "ds-gpio"
+    "ds-llrp"
+)
+
+## Device Service Descriptions
+declare -A deviceServiceDesc=(
+    [ds-bacnet]="Include BACnet Device Service"
+    [ds-camera]="Include Camera Device Service"
+    [ds-grove]="Include Grove Device Service (valid only: arm64)"
+    [ds-modbus]="Include ModBus Device Service"
+    [ds-mqtt]="Include MQTT Device Service"
+    [ds-rest]="Include REST Device Service"
+    [ds-snmp]="Include SNMP Device Service"
+    [ds-virtual]="Include Virtual Device Service"
+    [ds-coap]="Include CoAP Device Service"
+    [ds-gpio]="Include GPIO Device Service"
+    [ds-llrp]="Include RFID LLRP Device Service"
+
+)
```
You can use a nifty bash trick to reduce the amount of duplicated code.

Use `for index in "${!deviceServiceDesc[@]}";` to loop through the keys in the `deviceServiceDesc` associative array. This will allow you to remove the `deviceServices` array altogether.

Rinse and repeat for all similar code blocks.

shantanoo-desai commented on compose-builder/tui-generator.sh:
```diff
@@ -0,0 +1,299 @@
+SELECTED_APPSERVICES=()
+SELECTED_BUS=()
+SELECTED_OTHERS=()
+
```
One concern I have is that `tput` will mess up the terminal layout once the script exits the Whiptail TUI. This is because LINES and COLUMNS are generally available in a bash session and changing them might affect the spacing. I initially had some weird layout of my terminal after adding these variables but was able to reset it back using `tput clear`

shantanoo-desai replied:
based on the current changes @ajcasagrande can you please check to see if `tput` is changing some layout and spacing in your terminal after the script is executed?

lenny-intel reviewed (approved):
LGTM
